nebula-logger/core/main/plugin-framework/classes/LoggerPlugin.cls

```diff
@@ -0,0 +1,134 @@
+//------------------------------------------------------------------------------------------------//
+// This file is part of the Nebula Logger project, released under the MIT License.                //
+// See LICENSE file or go to https://github.com/jongpie/NebulaLogger for full license details.    //
+//------------------------------------------------------------------------------------------------//
+
+/**
+ * @group Plugin Framework
+ * @description The core of the plugin framework, used to create custom Apex & Flow plugins for `LoggerSObjectHandler` and `LogBatchPurger`
+ *              based on configurations stored in the custom metadata type `LoggerPlugin__mdt`
+ */
+public without sharing class LoggerPlugin {
+    private static final Map<String, LoggerPlugin__mdt> DEVELOPER_NAME_TO_RECORD = loadRecords();
+
+    /**
+     * @description Interface used to create plugins that can be used within Logger's batch job `LogBatchPurger`
+     */
+    public interface Batchable {
+        void start(LoggerPlugin__mdt configuration, LogBatchPurger.BatchPurgerInput input);
+        void execute(LoggerPlugin__mdt configuration, LogBatchPurger.BatchPurgerInput input, List<SObject> loggerRecords);
+        void finish(LoggerPlugin__mdt configuration, LogBatchPurger.BatchPurgerInput input);
+    }
+
+    /**
+     * @description Interface used to create plugins that can be used within Logger's trigger handler framework `LoggerSObjectHandler`
+     */
+    public interface Triggerable {
+        void execute(LoggerPlugin__mdt configuration, LoggerSObjectHandler.SObjectHandlerInput input);
+    }
+
+    public static List<LoggerPlugin__mdt> getBatchPurgerPluginConfigurations() {
+        List<PluginConfigurationSorter> matchingPluginConfigurationSorters = new List<PluginConfigurationSorter>();
+        for (LoggerPlugin__mdt pluginConfiguration : DEVELOPER_NAME_TO_RECORD.values()) {
+            if (pluginConfiguration.BatchPurgerApexClass__c != null || pluginConfiguration.BatchPurgerFlowName__c != null) {
+                matchingPluginConfigurationSorters.add(
+                    new PluginConfigurationSorter(pluginConfiguration)
+                        .sortBy(Schema.LoggerPlugin__mdt.BatchPurgerExecutionOrder__c)
+                        .sortBy(Schema.LoggerPlugin__mdt.DeveloperName)
+                );
+            }
+        }
+        matchingPluginConfigurationSorters.sort();
+        List<LoggerPlugin__mdt> matchingPluginConfigurations = new List<LoggerPlugin__mdt>();
+        for (PluginConfigurationSorter sorter : matchingPluginConfigurationSorters) {
+            matchingPluginConfigurations.add(sorter.pluginConfiguration);
+        }
+        return matchingPluginConfigurations;
+    }
+
+    public static List<LoggerPlugin__mdt> getSObjectHandlerPluginConfigurations() {
```

Comment:
this method, and `getBatchPurgerPluginConfigurations()` are so similar that I wonder if there's room for a shared method that takes in the flow/apex `Schema.SObjectField` tokens and returns the sorted plugins based on that 🤔

Reply:
Hmmm, yeah, I like that idea - that would reduce long-term maintenance/changes to the `LoggerPlugin` class + makes it easier elsewhere in the codebase to do things dynamically, to use filter on a list of fields, etc. I think I'm going to go with something like this new method signature:

`public static List<LoggerPlugin__mdt> getFilteredPluginConfigurations(List<Schema.SObjectField> populatedFilterFields, Schema.SObjectField sortByField)`

And then to replace the method `getSObjectHandlerPluginConfigurations()`, the new method will be called like this to get the same results:

```java
LoggerPlugin.getFilteredPluginConfigurations(
  new List<Schema.SObjectField>{ Schema.LoggerPlugin__mdt.SObjectHandlerApexClass__c, Schema.LoggerPlugin__mdt.SObjectHandlerFlowName__c },
  Schema.LoggerPlugin__mdt.SObjectHandlerExecutionOrder__c
);
```

Similarly, `LoggerPlugin.getBatchPurgerPluginConfigurations()` (used in `LogBatchPurger`) will also be replaced with: 
```java
this.pluginConfigurations = LoggerPlugin.getFilteredPluginConfigurations(
  new List<Schema.SObjectField>{ Schema.LoggerPlugin__mdt.BatchPurgerApexClass__c, Schema.LoggerPlugin__mdt.BatchPurgerFlowName__c },
  Schema.LoggerPlugin__mdt.BatchPurgerExecutionOrder__c
);
```

I also like that this change then delegates the responsibility of deciding "which fields to use" within `LogBatchPurger` and `LoggerSObjectHandler` - I feel like they should handle that logic, `LoggerPlugin` class shouldn't know that much about the usage of the CMDT records.

Reply:
🎉🎉